### DISCUSSION TITLE
CSP动态规划入门

### DISCUSSION THREAD

**Discussion by linzy666** [CSGrandeur/s-1problem1day1ac#490] [Algorithm]
### 动态规划（Dynamic Programming， DP）

动态规划是一种通过把原问题分解为相对简单的子问题的方法求解复杂问题的方法，与分治算法类似，都是通过组合子问题的解来求解原问题。但两者还有些差别，分治算法将问题划分为互不相交的子问题，通过递归的求解子问题，然后组合起来,但子问题和子子问题等上被重复计算了很多次。而动态规划应用与子问题重叠的情况，即不同的子问题具有公共的子子问题且具有记忆性，**通过填写表把所有已经解决的子问题答案纪录下来，在新问题里需要用到的子问题可以直接提取，避免了重复计算，从而节约了时间，所以在问题满足最优性原理之后，用动态规划解决问题的核心就在于填表，表填写完毕，最优解也就找到**。

### 引入

通过几个例题分别阐述动态规划的一些性质

###### 求斐波那契数列第k位的大小

在学习dp之前我们可能会用递归求解的方式来实现，但弊端是会出现重复调用同一个数的情况，如下图，重复出现了3、2 、1这些数字的计算，这样在位数稍微大点的时候就会非常的费时费力。我们可以来分析一下他的复杂度每个f[x]都要产生两个分支，新的f[x - 1] 和      f[x - 2]又要继续产生分支，所以对于调用的f[n]，总共会产生2^n个分支

![csp1](https://user-images.githubusercontent.com/90919399/184165466-ef81315b-3248-4693-97d4-9ae0a09b0bec.png)

```c++
#include<bits/stdc++.h>
using namespace std;
long long fib(int x)  //通过递归调求解子状态以此来求解该值
{
    if (x <= 0)
    {
        return 0;
    }
    if (x == 1) return 1;
    return fib(x - 1) + fib(x - 2);
}
int main()
{
    int k;
    scanf("%d",&k);
    printf("%lld",fib(k));
}
```

那怎么来优化前面那个算法，可能你会有一种想法是将计算过的值存储起来，每次计算的时候如果该值已经被计算了，则直接调用就好。

还有没有别的方法，我们知道斐波那契数列第i个数等于第i - 1个数加第i - 2个数，一直在用同一种计算方式重复计算。我们可不可以从小到大推， 依此构造最后求得完整解，最后的时间复杂度为O(n)。如果你理解了这个方法，恭喜你，你学会了最简单的dp

```c++
#include<bits/stdc++.h>
using namespace std;
long long dp[105];
long long fib(int x)
{
    if (x <= 0)
    {
        return 0;
    }
    if (x == 1) return 1;
    return fib (x - 1) + fib (x - 2);
}
int main()
{
    dp[0] = 0, dp[1] = 1;
    int k;
    scanf("%d",&k);
    for (int i = 2; i <= k; i++) dp[i] = dp[i - 1] + dp[i - 2];
    printf("%lld",dp[k]);
}
```

###### 无后效性：动态规划要求已经求解的子问题不受后续阶段的影响。（换言之，动态规划对状态空间的遍历构成一张有向无环图，遍历顺序就是该有向无环图的一个拓扑序。）

通过上面的例子，可以更好的理解这段话，第i位数字，即使计算了他后面一百位、一千位甚至无穷，但他的大小还是只由前两个数决定。

[P1216](https://www.luogu.com.cn/problem/P1216)

###### 题意：给定一个数字金字塔，让你找到一条路径从最高点走到最低点权值和最大

```c++
        7 
      3   8 
    8   1   0 
  2   7   4   4 
4   5   2   6   5 
```

这组数据很小我们通过观察会发现7 -> 3 -> 8 -> 7 -> 5的路径最大，但是如果他行的数量扩大到1000我们就没办法靠瞪眼法找出来了，让机器暴力跑出来也不太现实。你可能会很苦恼这个问题该怎么求解。别急这可是1994年 IOI的题目，当年在动态规划这个概念提出来不久时，这类问题可是非常难的。

我们换种分析方式，别从上往下走，改成从下往上走，当前这个点从其左下方和右下方传递过来，那么要最后权值最大，是不是要这个传递过来的值是max(l, r); 那么我们只要每一行计算完顺便记录他的值，就能慢慢一层一层推到顶点，达到最大值。

我们可以定义一个F[i] [j]表示从底部到第(i，j)产生的最大权值，有F[i+1] [j] 和 F[i+1] [j + 1]转移过来

```c++
#include <bits/stdc++.h>
using namespace std;
int a[1005][1005], f[1005][1005];
int main()
{
    int n;
    scanf("%d",&n);
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= i; j++)
        {
            scanf("%d",&a[i][j]);
            f[i][j]=a[i][j];
        }
    }
    for (int i = n - 1; i >= 1; i--)
    {
        for (int j = 1; j <= i; j++)
        {
            f[i][j] = max(f[i + 1][j], f[i + 1][j + 1]) + a[i][j];
        }
    }
    printf("%d",f[1][1]);
    return 0;
}
```

###### 最优子结构性质，很多情况下，动态规划用于解决最优化问题。此时，下一阶段的最优解应该能够由前面各阶段子问题的最优解导出。

通过上面的例子，我们也发现要当前值最大那子状态过来的那个值一定要子状态里面的最大。

#### 一个模型，三个特征：

（1）一个模型：

**“一个模型”指的是动态规划适合解决问题的模型。我把这个模型定义为“多阶段决策最优解模型”。**

具体来说，我们一般是用动态规划来解决最优问题。而解决问题的过程，需要经历多个决策阶段。每个决策阶段都对应一组状态。然后我们寻找一组决策序列，经过这组决策序列，能够产生最终期望求解的最优值。从解答树上看，决策是解答树产生一个分岔的描述，而最优解代表的就是一组决策序列，也就是从解答树起点到终点的一条路径。

（2）三个特征：

**①最优子结构：**

最优子结构指的是，问题的最优解包含子问题的最优解，反过来说就是，我们可以通过子问题的最优解，推导出问题的最优解。如果我们把最优子结构，对应到我们前面定义的动态规划问题模型上，那我们也可以理解为，后面阶段的状态可以通过前面状态推导出来。

最优子结构也说明了，问题的最优解不应该来源于子问题的非最优解（本来这样就根本做不到递推）、半成品解（还没更新完）。如果最优解的子决策序列对子问题来说不是最优的，那么说明“子问题”的定义仍太过笼统、不够具体，应该增加约束、上升维度。（不一定非要按照题目所问来确定问题所求，恰当地“化简为繁”，有利于找到切入方向）。

**②无后效性：**

无后效性，有两层含义，第一层含义是，在推导后面阶段状态的时候，我们只关心前面阶段的状态值，不关心这个状态是怎么一步步推导出来的。第二层含义是，某阶段状态一旦确定，就不受之后阶段的决策影响。无后效性是一个非常“宽松”的要求。只要满足前面提到的动态规划问题模型，其实基本上都会满足无后效性。

**③重复子问题：**

用一句话概括就是： 不同的决策序列，到达某个相同的阶段时，可能会产生重复的状态。

##### 总的来说，动态规划的特点可总结为两方面：

1) 问题具有最优子结构性质。如果问题的最优解所包含的 子问题的解也是最优的，我们就称该问题具有最优子结 构性质。

2) 无后效性。当前的若干个状态值一旦确定，则此后过程的演变就只和这若干个状态的值有关，和之前是采取哪种手段或经过哪条路径演变到当前的这若干个状态，没有关系。

### 动态规划步骤：

**（1）将原问题分解为子问题——**

根据具体题意以及约束条件来分解。子问题在形式上通常仍与原问题相同，只是问题规模变小了。

**（2）确定状态——**

很明显，动态规划与一般搜索技术最大不同的地方就是记录了已求解过的问题的结果。这里包含了两个方面的内容 ：子问题的记录和子问题结果的记录。其中，子问题的记录是最重要，也是最为复杂的，它就是通常我们所说的状态表示。通常我们用一个数、一组数或一个向量来实现状态表示。但无论采取什么方法，从动态规划的原理来看，状态表示要满足两个要求：正确、合理描述子问题和描述的子问题满足最优子结构性质；从算法实现角度来看，状态表示必须能够用基本数据结构实现并且能满足空间要求。我们得需要一些变量来保存子问题的结果，一般是用**数组**来保存。动态规划的时候需要开一个数组，每个数组的每个元素 f[i] 或 f\[i][j] 所代表的含义就是一个状态下的计算结果（描述状态需要几个变量就开几维的数组）。

**（3）确定初始状态和边界条件——**

动态规划必须有一个起点，在进行**初始化**后，才能开始状态之间的转移。

**（4）状态转移方程——**

即从已知状态迁移到某一未知状态的方程式，**转移方程是对旧状态与新状态之间的逻辑关系的描述**。

##### ”状态“ ”阶段“ 和 “决策” 是构成动态规划算法的三要素，而“子问题重叠性“ ”无后效性“ 和 “最优子结构性质” 是问题能用动态规划求解的三个基本条件。

因此，我们一般只需要定义出DP的计算过程，就可以编程实现了。 这个计算过程就被称为“状态转移方程”。

###### 问题描述：最长上升子序列。 给定一个长度为N 的数列，求数列单调递增的子序列的长度最长是多少。 例如 1 3 2 5 7 最长上升子序列可以是 1 3 5 7 也可以是 1 2 5 7

状态表示   F[i] 表示以 A[i]为结尾的“最长上升子序列”的长度

阶段划分 子序列的结尾位置 （数列A中的位置，从前到后）

转移方程 F[i] = max {F[j] + 1}  (0 <= j < i  && A[j] < A[i])

###### 问题描述：最长公共子序列。给定两个长度分别为N 和 M的字符串A和B， 求既是A的子序列又是B的子序列的字符串长度最长是多少。

状态表示 F[i, j] 表示前缀子串A[1 ~ i] 与 B[1 ~ j] 的 “最长公共子序列”的长度

阶段划分 已经处理的前缀长度 (两个字符串中的位置，即一个二维坐标)

转移方程 F[i, j] = max（F[i-1, j], F[i, j - 1], F[i-1,j-1] + 1(if ( A[i] == B[j] )）

LCS问题是比较复杂的问题，看前面的转移方程你可能不知道是啥。那换个理解方式，有两个串，分别是A串和B串，要求他们的的lcs是不是至少要让串A的每个字母都匹对一次串B的每个字母（可以想想你自己人脑模拟的过程），那么我们已经直到要一一匹对字母，那接下去的问题是，匹对成功值得怎么变化，匹对不成功值得怎么变化。

设子串为Z

###### 1、如果Ax = By 则 Zk = Ax = By 且 Zk-1 是 Ax-1和By-1的一个LCS 

当你现在是A串的x位和B串的y位匹对时，那么前面的串你肯定已经匹对好了，如果Ax = By 那么就是在前面匹对好的串加上一个Zk

###### 3、如果Ax != By 且Zk != By， 则Z是A和By-1的一个LCS

我们看看图片

![csp2](https://user-images.githubusercontent.com/90919399/184165681-bcbf3d9c-1bee-4fe4-98f7-160ed5f2d25a.png)
![csp3](https://user-images.githubusercontent.com/90919399/184165706-97f19c33-153c-47a3-b92f-eb9bf3326acc.png)
![csp4](https://user-images.githubusercontent.com/90919399/184165739-9f455f54-ce3c-433d-8c3d-e7e52eca26f2.png)
![csp5](https://user-images.githubusercontent.com/90919399/184165788-dca43b9d-9c6b-4dda-b563-2062878156cb.png)
![csp6](https://user-images.githubusercontent.com/90919399/184165805-39f7b072-4124-4772-bacb-437308d72b8f.png)
![csp7](https://user-images.githubusercontent.com/90919399/184165834-f458d98f-f9b0-42fd-b492-4c86e4d32248.png)
![csp8](https://user-images.githubusercontent.com/90919399/184165895-b0b381ce-1613-4f9f-94dc-48adf895701c.png)
![csp9](https://user-images.githubusercontent.com/90919399/184165899-06bedd6f-1385-44c7-98d1-19c6f1299336.png)
![csp10](https://user-images.githubusercontent.com/90919399/184165984-acc1a005-a174-4aca-858c-b06cb0c2bc42.png)
![csp11](https://user-images.githubusercontent.com/90919399/184166024-fd75ed6c-220f-47a2-ad2a-55d95ede279a.png)
![csp12](https://user-images.githubusercontent.com/90919399/184166044-a2be0ede-1813-44ec-9b84-3debed1a5f9f.png)
![csp13](https://user-images.githubusercontent.com/90919399/184166056-9b7e9c71-c8f2-4c81-931e-73f36a7d6a95.png)

以此类推

![csp14](https://user-images.githubusercontent.com/90919399/184166683-25588698-0cfd-4bb9-9162-bbb8bfef34d3.png)

### 区间DP

我们介绍的线性DP 一般从初态开始， 沿着阶段的扩张向某个方向递推，直至计算出目标状态。区间DP也属于线性DP中的一种，它以“区间长度”作为DP的“阶段”，使用两个坐标（区间的左、右端点）描述每个维度。我们直接上例题

###### 有N堆石子排成一排，其中第i堆石子的重量位Ai。每次可以选择其中相邻的两堆石子合并成一堆，形成的新石子堆的重量以及消耗的体力都是两堆石子的重量之和。求把全部N堆石子合成一堆最少需要消耗多少体力。

我们先来理解一下题目， 每次只能选择相邻的两堆石子合并成一堆，这意味着若第l堆石子和第r堆石子能合并成一堆，那么l ~ r之间的每堆石子已经被合并了，这样l 和 r才可能相邻。 因此，在任意时刻，任意一堆石子均可以用一个闭区间[l, r]来描述，表示这堆石子是由最初的第l ~ r堆石子合并而成的。那这堆石子的由来是不是又是由子两堆石子合并而来的，我们可以用一个k表示，这堆石子由第l ~ k 堆石子 和 k + 1 ~ r堆石子合并而来。

那么要让 l ~ r重量最小是不是就把问题落到了找一个最合适的k，自然的，应该把区间长度len作为DP的阶段，区间长度可以由两个端点表示出来。

###### 状态转移方程： F[l, r] = min{F[l, r] + F[k + 1, r]}  (l <= k < r) + sum _(l ~ r) 

```c++
#include<bits/stdc++.h>
using namespace std;
const int maxn=505;
#define INF 1e8
int n,dp[maxn][maxn],b[maxn],sum[maxn];
int main()
{
    scanf("%d",&n);
    for(int i=1;i<=n;i++) cin>>b[i];
    for(int i=1;i<=n;i++)
    {
        for(int j=1;j<=n;j++)dp[i][j]=INF;
    }
    for(int i=1;i<=n;i++)
    {
        dp[i][i]=0;
    }
    for(int i=1;i<=n;i++) sum[i]+=sum[i-1]+b[i];
    for(int len = 2; len <= n; len++) //区间长度 
    {
        for(int l=1;l<=n-len+1;l++) //区间起点 
        {
            int r = l + len - 1;         
            for(int k = l; k < r; k++)
            {
                dp[l][r] = min(dp[l][r],dp[l][k]+dp[k+1][r]+sum[r]-sum[l-1]);
            }
        }
    } 
    printf("%d",dp[1][n]);
}
```

区间DP的基本结构总结起来就是 O(n ^ 3)的时间复杂度

```c++
    for(int len = 2; len <= n; len++) //区间长度 
    {
        for(int l=1;l<=n-len+1;l++) //区间起点 
        {
            int r = l + len - 1;         
            for(int k = l; k < r; k++)
            {
                dp[l][r] = min(dp[l][r],dp[l][k]+dp[k+1][r]+sum[r]-sum[l-1]);
            }
        }
    } 
```

###### 给定两种括号 （） [] 当出现一个右括号找得到一个相应的左括号位匹配成功，给出一个只有括号做成的字符串，问最多有多少个括号满足要求。

还是来分析一下，如果第l个括号和第r个括号是匹配的，那么l + 1~ r - 1匹配成功的最大字符是不是可以直接加上这个 l 括号和 r括号。既然我们可以让l括号和r括号匹配 能不能让l括号和r括号不匹配，让他们各自去与 l +1 ~ r - 1的子括号匹配，最后取个最大值就为答案。

###### 然后与不匹配的方式的最大值进行比较dp[i] [j]=max(dp[i] [j],dp[i] [k]+dp[k] [j]);

```c++
#include <cstdio>
#include <queue>
#include <cstring>
#include <algorithm>
using namespace std;
 
typedef long long ll;
const int maxn = 105;

char s[maxn];
int dp[maxn][maxn];
 
int main()
{
    while(~scanf("%s",s+1)&&s[1]!='e')
    {
        int n=strlen(s+1);
        memset(dp, 0, sizeof(dp));
        for(int len=2;len<=n;len++)
        for(int i=1;i<=n;i++)
        {
            int j=i+len-1;
            if(j>n) break;
            if(s[i]=='('&&s[j]==')'||s[i]=='['&&s[j]==']')
            {
                dp[i][j]=dp[i+1][j-1]+2;
            }
            for(int k=i;k<j;k++)
            {
                dp[i][j]=max(dp[i][j],dp[i][k]+dp[k][j]);
            }
        }
        printf("%d\n",dp[1][n]);
    }
    return 0;
}

```

### 树形DP

给定一棵有N个节点的树 （通常是无根树， 也就是有N - 1条无向边），我们可以任选一个节点为根节点，从而定义出每个节点的深度和每棵子树的根。在树上进行一个动态规划算法时，一般就以节点从深到浅（子树从小到大）的顺序作为DP的“阶段”。DP的状态表示中，第一维通常是结点编号（代表以该节点位根的子树）。大多数时候我们采用递归的方式实现树形动态规划。对于每个节点x,先递归在它的每个子节点上进行DP，在回溯时，从子节点向节点x进行状态转移

######  某大学有N个职员，编号为1 - N。他们之间有从属关系，也就是说他们的关系就像一棵以校长为根的树，父结点就是子结点的直接上司。现在有个舞会，舞会每邀请来一个职员都会增加一定的快乐指数Hi。但是，如果某个职员的直接上司来参加舞会了，那么这个职员就无论如何也不肯来参加舞会了。   问：怎样邀请职员，能使快乐指数的总和最大。

分析可以直到，一个人去了那么他的下属一定不去，一个人不去他的下属可以去也可以不去。我们就以节点编号为DP状态第一维。以为一名职员是否愿意参加只跟他的直接上司是否参加有关，所以我们在每棵子树递归完成时，保留两个“代表信息” ——根节点参加时整棵子树的最大快乐指数和，以及根节点不参加时整棵子树的最大快乐指数和，就可以满足“最优子结构”性质了

###### 状态转移方程 F[x, 0] = x所有子节点分别的max(F[s, 0], F[s, 1])的和

```c++
const int N = 6005;
int dp[N][2]; //表示第i位员工是否参加的集合 属性：快乐指数最大
int r[N];
bool father[N];
vector<int> g[N];
int n;

void dfs(int u)
{
	dp[u][1] = r[u];

	for(auto v : g[u])
	{
		dfs(x); //先搜索求解f[v][0/1]
		dp[u][1] += dp[v][0]; //下属只能不去
		dp[u][0] += max(dp[v][0], dp[v][1]); //下属去不去都行
	}
}

int main()
{
	scanf("%d", &n);
	for(int i = 1; i <= n; i ++)
		scanf("%d", &r[i]);

	for(int i = 1; i < n; i ++)
	{
		int v, u;
		scanf("%d%d", &v, &u);
		g[u].pb(v);
		father[v] = true;
	}

	int root = -1;
	for(int i = 1; i <= n; i ++) 
		if(!father[i]) //没有父亲的节点就是根节点
			{root = i; break;}

	dfs(root);

	printf("%d\n", max(dp[root][0], dp[root][1]));
}
```

## 状态压缩DP

状压dp（状态压缩动态规划）的特点一般是**规模比较小**，n一般小于15，而且每个元素**一般只有两种形态**。比某个开关打不打开，某个物品拿或不拿，某个格子放不放棋子，等等。

状压dp是一类比较难理解的dp。在讲状压dp之前，我们应该清楚所有的dp是解决多阶段决策最优化问题的一种思想方法，请注意多阶段这三个字，经过前面基础DP的学习，可以发现如何定义状态是解决动态规划最重要的一步，状态的定义也就决定了相当于阶段的划分。

动态规划多阶段一个重要的特性就是无后效性。无后效性就是值对于某个给定的阶段状态，它以前各阶段的状态无法直接影响它未来的发展，而只能通过当前的这个状态。换句话说影响当前阶段状态只可能是前一阶段的状态。可以看出如何定义状态是至关重要的，因为状态决定了阶段的划分，阶段的划分保证了无后效性。

我们拿01背包举例（当然，这不是典型的状压DP问题）：设dp[x]为选取方式为状态x时的总价值。

我们仔细观察就会发现，每件物品有放与不放两种选择；假设我们有5件物品的时候，用1和0代表放和不放，如果这5件物品都不放的话，那就是00000； 如果这5件物品都放的话，那就是11111。**可见我们可以用二进制表示所有物品的放与不放的情况**。放的状态只能从不放的状态转移过来，所以dp[10000]只能从dp[00000] + value[1] 转移过来，dp[11000]可以从dp[01000] + value[1]或者dp[10000] + value[2]转移过来.........。

掌握状态压缩DP的基本功是充分了解位运算，以下是常用的位操作：

**注：在涉及到位运算时，一定要注意位运算的优先级。该加的括号一定要加**

x | 1 -1：把最后一位变成0

x & (-x)：得到最低位的1（其后都为0）

x & (1<<k -1)：取右侧k位

a&(a<<1)：a中没有两个1相邻

a^b=a+b：a与b的二进制任何一位都不可以同时为1，等价于a&b=0

x &= ( x - 1 )： 把最右侧的1置0，循环此操作可统计有几个1

x |= (x + 1)：把最右侧的0置1

x & (x + 1)：把右边连续的1变成0

x | (x - 1)：把右边连续的0变成1

 （加一或减一，最低位一定会变化。加一时，右侧连续的1会变0，直到遇见0；减一时，右侧连续的0会变1，直到遇见1）

state &= ~ST：状态中去掉ST状态

state & ST == ST：判断当前状态有没有包含ST状态

state |= ST：状态中添加 ST状态

状压DP的题目较为灵活，大多数时候，我们可以参考下面的模式——

```cpp
int n;
int up = 1 << n;//总状态数。
//枚举题给的所有集合。按照状态转移的顺序，一般从小编号到大编号。
for(int i = 1; i <= m; ++ i){
    //枚举当前集合中的状态。
    for(int now = 0; now < up; ++ now){
        //通常要进行两次判断：1.状态now本身是否合法（满足题目约束条件）；2.状态now对于当前集合是否是合法状态
        if(状态合法){
            for(int st = 0; st < up; ++ st){
                //进行两次判断：1.状态st本身是否合法（满足题目约束条件）；2.状态st是否可以向now转移。
                if(状态合法){
                    //列写状态转移方程。
                }
            }
        }
    }
}
int ans=0;
for() //通过for循环把所有的合法值累加到结果
```

以一道经典题为例：[USACO06NOV\]Corn Fields G](https://www.luogu.com.cn/problem/P1879)
题目解释：农场主 John 新买了一块长方形的新牧场，这块牧场被划分成 M行 N 列 (1≤M≤12;1≤N≤12)，每一格都是一块正方形的土地。John打算在牧场上的某几格里种上美味的草，供他的奶牛们享用。遗憾的是，有些土地相当贫瘠，不能用来种草。并且，奶牛们喜欢独占一块草地的感觉，于是 John 不会选择两块相邻的土地，也就是说，没有哪两块草地有公共边。John 想知道，如果不考虑草地的总块数，那么，一共有多少种种植方案可供他选择？

我们拿01背包举例（当然，这不是典型的状压DP问题）：共用n件物品，每种物品仅一个，设dp[x]为选取方式为状态x时的总价值。

代码：

```cpp
long long dp[14][1<<12]; //dp[i][j]表示第i行的状态为j的1~i行的总方案数
int check1(int x) //判断状态x中，是否存在两个1相邻
{
    return x&(x<<1);
}
int check2(int x,int y)
{
    return x&y; //判断状态x和y是否在同一位置都是1
}
int main()
{
    int n,m;
    cin>>n>>m;
    int up=(1<<m);
    dp[0][0]=1;
    for(int i=1;i<=n;++i)
    {
        char ch;
        int cur=0;
        for(int j=0;j<m;++j) //将输入的每一行的01串，变成十进制数字cur，表示每一行的状态集合
        {
            cin>>ch;
            if(ch=='1') 
                cur+=(1<<j);
        }
        for(int st=0;st<up;++st)
        {
            if((st&cur)!=st) continue; //如果st不是cur的子状态，则跳过
            if(check1(st)) continue;
            for(int k=0;k<up;++k)
            {
                if(check2(st,k)) continue;
                dp[i][st]+=dp[i-1][k];
                dp[i][st]%=mod;
            }
        }
    }
    long long ans=0;
    for(int i=0;i<up;++i)
        ans+=dp[n][i],ans%=mod;
    cout<<ans;
}
```

虽然状压DP的状态转移推导并不困难，但是关于状态合法的分析判断有时候并不容易，常常会因为漏掉情况而WA掉。如果比较难去知道每个状态都是从那些合状态转移而来的，或者状态的特殊情况过多，我们可以考虑自后向前，将其转换为较为直观的**记忆化搜索**的形式（比如说[CCF-CSP 202109-4](http://118.190.20.162/view.page?gpid=T132)）。以下代码是上述问题的记忆化搜索版。

```cpp
long long dp[14][1<<12];
int n,m,up,S[14];
int check1(int x)
{
    return x&(x<<1);
}
int check2(int x,int y)
{
    return x&y;
}
long long dfs(int p,int now)
{
    if(p==n) return 1;
    if(dp[p][now]) return dp[p][now];
    for(int st=0;st<up;++st)
    {
        if((st&S[p+1])!=st) continue;
        if(check1(st)) continue;
        if(check2(st,now)) continue;
        dp[p][now]+=dfs(p+1,st);
        dp[p][now]%=mod;
    }
    return dp[p][now];
}
int main()
{
    cin>>n>>m;
    up=(1<<m);
    for(int i=1;i<=n;++i)
    {
        char ch;
        for(int j=0;j<m;++j)
        {
            cin>>ch;
            if(ch=='1') 
                S[i]+=(1<<j);
        }
    }
    cout<<dfs(0,0);
}
```

经典题：[SCOI2005\] 互不侵犯](https://www.luogu.com.cn/problem/P1896)

思路：状态 $dp[i][j][k]$ 表示1~i的 j 个国王的摆放方案，其中第 i 行的摆放方式为k。状态转移方程:$f[i][j][S]=\sumf[i-1][k][T]$，其中 $j>=k$ 且  $j-k=S中的1的数量$  。

思考：“N皇后”问题可以运用状态DP吗？（“皇后”可沿八方向直线攻击）

“N皇后”问题虽然可以运用状态压缩来记录棋盘状态，但是无法采取动态规划，因为这个问题不具备“无后效性”，因为所有新放置的皇后，都要受到所有之前放置的皇后的限制。

**Comment by FM3676** [comment]
**_### 很好！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！受益颇多，爱来自中国_**

**Comment by FM3676** [comment]
**_### 很好！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！受益颇多，爱来自中国_**
